kythe/go/extractors/config/smoke/smoke.go

```diff
@@ -72,12 +87,20 @@ type harness struct {
 //
 // An extraction config can be optionally read from a specified file.  The
 // format follows kythe.proto.ExtractionConfiguration.
-func NewGitTestingHarness(configPath string) Tester {
-	return harness{
+//
+// Finally, the testing harness can optionally try to index the extraction
+// output.
+func NewGitTestingHarness(configPath string, doIndex bool) Tester {
```

Comment:
It might be worth breaking out a settings type here, e.g., 

```
type HarnessOptions struct {
	Indexer   Indexer
	Fetcher   Fetcher
	Extractor Extractor
}
```

This makes it easy to provide defaults:

```
func (o *HarnessOptions) extractor() Extractor {
	if o == nil || o.Extractor == nil {
		return config.DefaultExtractor{}
	}
	return o.Extractor
}
// etc.
```

And then the constructor has a simpler form:

```
func NewTestingHarness(configPath string, opts *HarnessOptions) Tester {
	h := &harness{
		extractor:  opts.extractor(),
		configPath: configPath,
		fetcher:    opts.fetcher(),
		indexer:    opts.indexer(),
	}
	return h.TestRepo
}
```

Then the user can do what they want without having to write a bunch of separate constructors:

```
	h := smoke.NewTestingHarness("github.com/google/kythe.git", &smoke.HarnessOptions{
		Fetcher: smoke.GitFetch,
		Indexer: indexers.Java,
	})
```

There's more in this than you might want to do in a single change, but I'm a little concerned by the proliferation of interfaces without much generality. Though even if we keep those, I think the options-type has potential to simplify some of this.

Reply:
Interesting.  Why not just use the HarnessOptions as the only type thing there, and stuff the methods on it?  (See commi)

Reply:
Yeah, that's also a viable approach. I often separate the types if the underlying type has mutable state, so that the options don't get mixed up with hidden details. But here it'd arguably be fine to export the settings directly, and just implement everything as methods as you say.

In that case, you could even maybe do away with the constructor,

```
// A Harness contains the settings necessary to test a repository.
// The ConfigPath is required, all the other fields are optional and use
// defaults as specified.
type Harness struct {
  ConfigPath string
   Extractor Extractor  // defaults to blah
  ...
}
```